src/certs.cc

```diff
@@ -0,0 +1,175 @@
+#include "certs.h"
+#include <functional>
+#include <stdexcept>
+#include <wincrypt.h>
+
+namespace WinExportCertificateAndKey {
+namespace {
+
+// Naive RAII cleanup helper.
+struct Cleanup {
+  std::function<void()> fn;
+  Cleanup(std::function<void()> fn) : fn(fn) {}
+  ~Cleanup() { fn(); }
+};
+
+// Throw an exception based on the last Windows error message.
+void ThrowWindowsError(const char* call) {
+  DWORD err = GetLastError();
+  CHAR err_msg_buf[128];
```

Comment:
128 might be pretty short for an error message. It's in the stack, maybe we can be a bit more generous? What would happen if FormatMessageA fails? [Looking at the docs](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-formatmessage) I don't really see what would happen, so maybe using FORMAT_MESSAGE_ALLOCATE_BUFFER and the freeing the memory with LocalFree?

Reply:
This is doing the Windows equivalent of `strerror()` – I'd be surprised if any single Windows error message had more than 128 characters. But yeah, worth thinking about, and it's stack space, so it's cheap. Added another 128 bytes in 924143efc0e4ee8cc7adcc569b5d77435757f3d3 🙂